code/modules/atmospherics/machinery/components/unary_devices/cryo.dm

```diff
@@ -11,34 +11,45 @@
 	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
 	pixel_y = 22
 	appearance_flags = KEEP_TOGETHER
+	var/mob/living/occupant
 
-/atom/movable/visual/cryo_occupant/Initialize()
+/atom/movable/visual/cryo_occupant/Initialize(mapload, obj/machinery/atmospherics/components/unary/cryo_cell/parent)
 	. = ..()
 	// Alpha masking
 	// It will follow this as the animation goes, but that's no problem as the "mask" icon state
 	// already accounts for this.
 	add_filter("alpha_mask", 1, list("type" = "alpha", "icon" = icon('icons/obj/cryogenics.dmi', "mask"), "y" = -22))
+	RegisterSignal(parent, COMSIG_MACHINERY_SET_OCCUPANT, .proc/on_set_occupant)
+	RegisterSignal(parent, COMSIG_CRYO_SET_ON, .proc/on_set_on)
 
-/atom/movable/visual/cryo_occupant/proc/on_occupant_enter(mob/living/occupant)
+/atom/movable/visual/cryo_occupant/proc/on_set_occupant(datum/source, mob/living/L)
+	SIGNAL_HANDLER
+
+	if(occupant)
+		vis_contents -= occupant
+		REMOVE_TRAIT(occupant, TRAIT_IMMOBILIZED, CRYO_TRAIT)
+		if(occupant.resting || HAS_TRAIT(occupant, TRAIT_FLOORED))
+			occupant.set_lying_down()
+	if(!L || !istype(L))
+		occupant = null
+		return
+
+	occupant = L
```

Comment:
Register to the new occupant's `COMSIG_PARENT_QDELETING` and handle the event of it being deleted so we don't keep references and cause a hard deletion.

Reply:
`machinery` already calls `set_occupant(null)` on destruction so this shouldn't be an issue, I guess?